cpp/src/arrow/compute/kernels/temporal_internal.h

```diff
@@ -40,15 +39,42 @@ using arrow_vendored::date::year_month_day;
 using arrow_vendored::date::zoned_time;
 using std::chrono::duration_cast;
 
+using ArrowTimeZone = std::variant<const time_zone*, const OffsetZone>;
```

Comment:
@pitrou I tried:
```cpp
using ArrowTimeZone = std::variant<const time_zone*, const OffsetZone>;
```
but it was a bit difficult to not have OffsetZone go out of scope. I think this is fine though/